apps/lockfile-explorer-web/src/parsing/lfxGraphLoader.ts

```diff
@@ -0,0 +1,363 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { LfxGraph, LockfileEntry, LockfileEntryFilter } from './LfxGraph';
+import { DependencyKind, LockfileDependency } from './LfxGraph';
+import { Path } from '@lifaon/path';
+
+enum PnpmLockfileVersion {
+  V6,
+  V5
+}
+
+export interface ILockfileImporterV6 {
+  dependencies?: {
+    [key in string]: {
+      specifier: string;
+      version: string;
+    };
+  };
+  devDependencies?: {
+    [key in string]: {
+      specifier: string;
+      version: string;
+    };
+  };
+}
+export interface ILockfileImporterV5 {
+  specifiers?: Record<string, string>;
+  dependencies?: Record<string, string>;
+  devDependencies?: Record<string, string>;
+}
+export interface ILockfilePackageType {
+  lockfileVersion: number | string;
+  importers?: {
+    [key in string]: ILockfileImporterV5 | ILockfileImporterV6;
+  };
+  packages?: {
+    [key in string]: {
+      resolution: {
+        integrity: string;
+      };
+      dependencies?: Record<string, string>;
+      peerDependencies?: Record<string, string>;
+      dev: boolean;
+    };
+  };
+}
+
+export interface ILockfileNode {
+  dependencies?: {
+    [key in string]: string;
+  };
+  devDependencies?: {
+    [key in string]: string;
+  };
+  peerDependencies?: {
+    [key in string]: string;
+  };
+  peerDependenciesMeta?: {
+    [key in string]: {
+      optional: boolean;
+    };
+  };
+  transitivePeerDependencies?: string[];
+}
+
+const packageEntryIdRegex: RegExp = new RegExp('/(.*)/([^/]+)$');
+
+function createLockfileDependency(
+  name: string,
+  version: string,
+  dependencyType: DependencyKind,
+  containingEntry: LockfileEntry,
+  node?: ILockfileNode
+): LockfileDependency {
+  const result: LockfileDependency = new LockfileDependency({
+    name,
+    version,
+    dependencyType,
+    containingEntry
+  });
+
+  if (result.version.startsWith('link:')) {
+    const relativePath = result.version.substring('link:'.length);
+    const rootRelativePath = new Path('.').relative(
+      new Path(containingEntry.packageJsonFolderPath).concat(relativePath)
+    );
+    if (!rootRelativePath) {
+      // eslint-disable-next-line no-console
+      console.error('No root relative path for dependency!', name);
```

Comment:
This isn't fatal?

Reply:
Agreed, a lot of this parsing seems to silently discard unexpected inputs. That's typical for a lot of  web apps, but how are we supposed to debug complex build tools algorithms if their foundation is hiding problems?

I can't fix it easily in a refactoring PR, though. This code will get another pass when we work PNPM 9 support.